File: openFrameworks_testing/src/bubble.cpp
//
//  bubble.cpp
//  openFrameworks_testing
//
//  Created by Kevin Nottberg on 2/2/18.
//

#include <stdio.h>
#include <math.h>
#include "bubble.h"


Bubble::Bubble( int radius, int mass, int ID, int resolution )
    :position(0, 0), velocity(0, 0), acceleration(0, 0), preExpansivePos(position.copy()),
        expansiveVelocity(0, 0), textField(100, 100, 800, 800)
{
    this->radius = radius;
    this->mass = mass;
    this->ID = ID;
    this->resolution = resolution;
    
    this->textField = TextField(50, 50, 400, 600);
    this->textField.setFieldPadding(10);
    this->textField.setBackgroundColor(255, 255, 255, 155);
}

void Bubble::setup() {
    
    //Initalize Position and Velocity vector for bubbles
    this->position.x_comp = rand() % (ofGetWindowWidth() - this->radius) + this->radius;
    this->position.y_comp = rand() % (ofGetWindowHeight() - this->radius) + this->radius;
    
    this->velocity.x_comp = rand() % 4;
    this->velocity.y_comp = rand() % 4;
    
    //Generate the current vert list for the bubble resolution
    for( int i = 0; i < this->resolution; i += 1) {
        float theta = (2.0 * 3.145926 * i)/this->resolution;
        float x = this->radius * cos(theta);
        float y = this->radius * sin(theta);
        this->vertList.push_back((x));
        this->vertList.push_back((y));
    }
    
    //Shine Vert list for bubble
    for( int i = 0; i < int( ceil(this->resolution/8) ); ++i ) {
        float theta = 2.0 * 3.145926 * (i/this->resolution);
        
        this->shineVertList.push_back(int( this->radius * cos(theta) ));
        this->shineVertList.push_back(int( this->radius * sin(theta) ));
    }
}

void Bubble::draw() {
    //Draw call here
    ofSetPolyMode(OF_POLY_WINDING_NONZERO);
    ofSetLineWidth(4);
    ofBeginShape();
    for(int i = 0; i < this->vertList.size(); i += 2) {
        ofVertex(this->position.x_comp + this->vertList[i],
                   this->position.y_comp + this->vertList[i + 1]);
        //cout << this->vertList[i] << endl;
    }
    ofEndShape();
    
    ofSetPolyMode(OF_POLY_WINDING_NONZERO);
    ofSetLineWidth(4);
    ofBeginShape();
    for(int i = 0; i < this->shineVertList.size(); i += 2) {
        ofVertex(this->position.x_comp + this->shineVertList[i],
                   this->position.y_comp + this->shineVertList[i + 1]);
    }
    ofEndShape();
}

void Bubble::expansiveDraw() {
    ofSetPolyMode(OF_POLY_WINDING_NONZERO);
    ofSetLineWidth(4);
    ofBeginShape();
    for( int i = 0; i < this->resolution; i += 1 ) {
        float theta = (2.0 * 3.145926 * i)/this->resolution;
        float x = this->radius * cos(theta);
        float y = this->radius * sin(theta);
        ofVertex(this->position.x_comp + x, this->position.y_comp + y);
    }
    ofEndShape();
}

void Bubble::notingDraw() {
    this->textField.draw();
}

void Bubble::update(RenderState state) {
    //Update the vector positioning
    if(state == RenderState::floating) {
        this->position.add(this->velocity);
        if(this->position.x_comp >= ofGetWindowWidth() - this->radius) {
            this->velocity.x_comp *= -1;
            this->position.x_comp = ofGetWindowWidth() - this->radius - 3;
        } else if (this->position.x_comp <= this->radius) {
            this->velocity.x_comp *= -1;
            this->position.x_comp = (this->radius + 3);
        } else if (this->position.y_comp >= ofGetWindowHeight() - this->radius) {
            this->velocity.y_comp *= -1;
            this->position.y_comp = ofGetWindowHeight() - this->radius - 3;
        } else if (this->position.y_comp <= this->radius) {
            this->velocity.y_comp *= -1;
            this->position.y_comp = this->radius + 3;
        }
        
        this->velocity.limit(3);
        this->position.x_comp = (int(this->position.x_comp));
        this->position.y_comp = (int(this->position.y_comp));
    } else if (state == RenderState::expanding) {
        this->position.add(this->expansiveVelocity);
    } else if (state == RenderState::closing) {
        this->position.sub(this->expansiveVelocity);
    }
}

void Bubble::setPosition(int x, int y) {
    this->position.x_comp = x;
    this->position.y_comp = y;
}


void Bubble::keyboardInput(int key) {
    this->textField.processInput(key);
}


void Bubble::setExpansiveVelocity(float x_comp, float y_comp) {
    this->expansiveVelocity.x_comp = x_comp;
    this->expansiveVelocity.y_comp = y_comp;
}

File: Floating_Note_OpenGL/Bubble.py
from simple_vector_library import Py_Vector
import numpy as np
import math
from random import *
import global_settings as gls

from OpenGL.GL import *
from OpenGL.GLU import *

class Bubble:
    def __init__(self, radius, mass, ID, resolution):
        #self.pos = Py_Vector(randint(radius, gls.SCREENWIDTH - radius), randint(radius, gls.SCREENHEIGHT - radius))
        self.pos = Py_Vector(0, 0)
        self.vel = Py_Vector(randint(-5, 5), randint(-4, 4))
        self.accel = Py_Vector(0, 0)

        self.resolution = resolution

        self.preExpansivePos = Py_Vector(0, 0)
        self.expansiveVelocity = Py_Vector(0, 0)

        self.radius = radius
        self.mass = mass

        self.ID = ID

        self.noteList = []

        self.vertList = []
        self.shineVertList = []
        self.calculateVerticies()

    '''
    def draw(self):
        glLineWidth(4)
        glBegin(GL_LINE_LOOP)
        for i in range(self.resolution):
            theta = 2.0 * 3.1415926 * i/self.resolution
            x = self.radius * math.cos(theta)
            y = self.radius * math.sin(theta)
            glVertex2f(self.pos.x_component + x, self.pos.y_component + y)
        glEnd()

        #Bubble shine
        glBegin(GL_LINE_LOOP)
        for i in range(math.ceil(self.resolution/8)):
            theta = 2.0 * 3.1415926 * i/(self.resolution)
            x = (self.radius * 0.75) * np.cos(theta)
            y = (self.radius * 0.75) * np.sin(theta)
            glVertex2f(self.pos.x_component + x, self.pos.y_component + y)
        glEnd()
    '''

    def draw(self):
        glLineWidth(5)
        glBegin(GL_LINE_LOOP)
        for i in range(0, len(self.vertList,), 2):
            glVertex2f(self.pos.x_component + self.vertList[i],
                        self.pos.y_component + self.vertList[i + 1])

        glEnd()

        glBegin(GL_LINE_LOOP)
        for i in range(0, len(self.shineVertList), 2):
            glVertex2f(self.pos.x_component + self.shineVertList[i],
                        self.pos.y_component + self.shineVertList[i + 1])
        glEnd()

    def calculateVerticies(self):
        for i in range(self.resolution):
            theta = 2.0 * 3.1415926 * (i/self.resolution)
            x = self.radius * math.cos(theta)
            y = self.radius * math.sin(theta)
            self.vertList.append(math.ceil(x))
            self.vertList.append(math.ceil(y))

        for i in range(math.ceil(self.resolution/8)):
            theta = 2.0 * 3.145926 * i/(self.resolution)
            x = (self.radius * 0.75) * math.cos(theta)
            y = (self.radius * 0.75) * math.sin(theta)
            self.shineVertList.append(math.ceil(x))
            self.shineVertList.append(math.ceil(y))


    def update(self, STATE):
        if(STATE == 'floating'):
            #this.vel.add(this.accel)
            self.pos.add(self.vel)
            if(self.pos.x_component >= gls.SCREENWIDTH - (self.radius + 3)):
                self.vel.x_component *= -1
                self.pos.x_component = (gls.SCREENWIDTH - (self.radius + 3))

            if(self.pos.x_component <= self.radius + 3):
                self.vel.x_component *= -1
                self.pos.x_component = (self.radius + 3)

            if(self.pos.y_component >= gls.SCREENHEIGHT - (self.radius+3)):
                self.vel.y_component *= -1
                self.pos.y_component = (gls.SCREENHEIGHT - (self.radius + 3))

            if(self.pos.y_component <= self.radius+3):
                self.vel.y_component *= -1
                self.pos.y_component = (self.radius + 3)
        elif(STATE == 'expansion'):
            self.pos.add(self.expansiveVelocity)
        elif(STATE == 'closing'):
            self.pos.sub(self.expansiveVelcoity)

        self.vel.limit(5)

    def expansiveDraw(self):
        glLineWidth(4)
        glBegin(GL_LINE_LOOP)
        for i in range(self.resolution):
            theta = 2.0 * 3.1415926 * i/self.resolution
            x = self.radius * math.cos(theta)
            y = self.radius * math.sin(theta)
            glVertex2f(self.pos.x_component + x, self.pos.y_component + y)
        glEnd()

        #Removing the shine as the bubbles are bouncing speeds up
        # the expansion animation
        '''
        #Bubble shine
        glBegin(GL_TRIANGLES)
        for i in range(math.ceil(self.resolution/8)):
            theta = 2.0 * 3.1415926 * i/(self.resolution)
            x = (self.radius * 0.75) * np.cos(theta)
            y = (self.radius * 0.75) * np.sin(theta)
            glVertex2f(self.pos.x_component + x, self.pos.y_component + y)
        glEnd()
        '''
